8300258: C2: vectorization fails on simple ByteBuffer loop

rwestrel opened the pull request:
The loop that doesn't vectorize is:

```
public static void testByteLong4(byte[] dest, long[] src, int start, int stop) {
    for (int i = start; i < stop; i++) {
        UNSAFE.putLongUnaligned(dest, 8 * i + baseOffset, src[i]);
    }
}
```

It's from a micro-benchmark in the panama
repo. `SuperWord::find_adjacent_refs() `prevents it from vectorizing
because it finds it cannot properly align the loop and, from the
comment in the code, that:

```
// Can't allow vectorization of unaligned memory accesses with the
// same type since it could be overlapped accesses to the same array.

```
The test for "same type" is implemented by looking at the memory
operation type which in this case is overly conservative as the loop
above is reading and writing with long loads/stores but from and to
arrays of different types that can't overlap. Actually, with such
mismatched accesses, it's also likely an incorrect test (reading and
writing could be to the same array with loads/stores that use
different operand size) eventhough I couldn't write a test case that
would trigger an incorrect execution.

As a fix, I propose implementing the "same type" test by looking at
memory aliases instead.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300258](https://bugs.openjdk.org/browse/JDK-8300258): C2: vectorization fails on simple ByteBuffer loop


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - Committer) ⚠️ Review applies to [f6820c45](https://git.openjdk.org/jdk/pull/12440/files/f6820c45e9648b2e4f0bf1b5529458e3b3ff3cc5)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [f6820c45](https://git.openjdk.org/jdk/pull/12440/files/f6820c45e9648b2e4f0bf1b5529458e3b3ff3cc5)


### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12440/head:pull/12440` \
`$ git checkout pull/12440`

Update a local copy of the PR: \
`$ git checkout pull/12440` \
`$ git pull https://git.openjdk.org/jdk pull/12440/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12440`

View PR using the GUI difftool: \
`$ git pr show -t 12440`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12440.diff">https://git.openjdk.org/jdk/pull/12440.diff</a>

</details>

eme64 commented:
@rwestrel is there a reason why you only replaced these two instances of `same_velt_type`?
The logic of "same type" is also used later on in the code of `SuperWord::find_adjacent_refs()`.
For example when we have `create_pack == false`, we remove all memops and packs of the "same type". Do you not want to change that too?

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -662,7 +662,7 @@ void SuperWord::find_adjacent_refs() {
       }
     } else {
-      if (same_velt_type(mem_ref, best_align_to_mem_ref)) {
+      if (_phase->C->get_alias_index(mem_ref->adr_type()) == _phase->C->get_alias_index(best_align_to_mem_ref->adr_type())) {
```
Would it not be nicer to wrap this in a separate function?
`same_memory_slice(mem_ref, best_align_to_mem_ref)`
You can then reuse it everywhere, makes it more readable.

You should also change the comments below, and elsewhere, as it is now not about "same type", but "same memory slice".

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java:
```diff
@@ -0,0 +1,184 @@
+ * @run driver compiler.c2.irTests.TestVectorizationMismatchedAccess
+ */
+
+public class TestVectorizationMismatchedAccess {
```
Could you have a much simpler example, just with two int-arrays, where one allows vectorization and one does not. We still want there to be some vectorization. Maybe like this, maybe you have to play around with it a bit:
```
// make sure arr1 and arr2 are in different slices
for (int i ...) {
  arr1[i] = arr1[i-2] + 5; // do not vectorize, cyclic dependency
  arr2[i] = arr2[i] + 5; // should vectorize
}
```

rwestrel replied:
I agree that we should have such a test if we don't have one already but does it belong to that bug? That bug changes the behavior with mismatched accesses so it feels strange to add an unrelated test.` testByteByte1() `is a mismatched access test that vectorizes and that's unaffected by my change. I could add another one that doesn't vectorize because of a dependency with a similar mismatched access. What do you think?

eme64 replied:
Well, my idea was this:
You added a new feature, where we now have a more fine-grained logic. Before, we distinguished by "velt type". Now we distinguish by "memory slice".
Before, this change, my `arr1` and `arr2` would have had the same "velt type", and if one is rejected from vectorization, so is the other. Now they are handled separately, right?

vnkozlov replied:
I agree with Emanuel here. We need negative tests here too - correctly reject vectorizations which may cause cyclic dependency bugs. For example, passing the same array as 2 arguments (as in `testByteByte1()`) but with different offset to create cyclic dependency.

eme64 replied:
I understand the code better now. I believe that two arrays of the same basic type never land in two separate memory sliced, but always the same. I discussed this with @rwestrel  a while ago. The reason is that EA does not give arrays their own slice, it determines the array to be NSR (not scalar replaceable) if it is ever used in a loop. That is of course a shame, and I hope we can work on this in the future (maybe have some analysis in SuperWord that refines the memory slices, and thus removes unnecessary dependencies). But as @rwestrel  says not the topic of this Bug.
The goal of this thread was to find an example where @rwestrel 's first commit made a worse decision than before. And I found another one after wards (https://github.com/openjdk/jdk/pull/12440#issuecomment-1425919095). @rwestrel now updated the fix, and this specific test should no longer fail.
@vnkozlov adding cyclic dependency tests before fixing https://github.com/openjdk/jdk/pull/12350 will be difficult. Maybe we need to fix both bugs first, and then once they are both integrated add more tests?

eme64 reviewed (changes requested):
@rwestrel is there a reason why you only replaced these two instances of same_velt_type?
The logic of "same type" is also used later on in the code of SuperWord::find_adjacent_refs().
For example when we have create_pack == false, we remove all memops and packs of the "same type". Do you not want to change that too?

For example here:
https://github.com/openjdk/jdk/blob/8f195ff236000d9c019f8beb2b13355083e211b5/src/hotspot/share/opto/superword.cpp#L710-L738

But maybe also this should now be adapted:
https://github.com/openjdk/jdk/blob/8f195ff236000d9c019f8beb2b13355083e211b5/src/hotspot/share/opto/superword.cpp#L1268-L1271

rwestrel commented:
> @rwestrel is there a reason why you only replaced these two instances of `same_velt_type`? The logic of "same type" is also used later on in the code of `SuperWord::find_adjacent_refs()`. For example when we have `create_pack == false`, we remove all memops and packs of the "same type". Do you not want to change that too?

They don't have the same purpose AFAIU. The remaining uses in `SuperWord::find_adjacent_refs()` are related to the heuristic to find the best alignment. There it makes sense that the actual size of the operation is what matters, I think. While the one I change deal with correctness in the face of possibly overlapping arrays. So, yes, It's on purpose.

eme64 commented:
> > @rwestrel is there a reason why you only replaced these two instances of `same_velt_type`? The logic of "same type" is also used later on in the code of `SuperWord::find_adjacent_refs()`. For example when we have `create_pack == false`, we remove all memops and packs of the "same type". Do you not want to change that too?
> 
> They don't have the same purpose AFAIU. The remaining uses in `SuperWord::find_adjacent_refs()` are related to the heuristic to find the best alignment. There it makes sense that the actual size of the operation is what matters, I think. While the one I change deal with correctness in the face of possibly overlapping arrays. So, yes, It's on purpose.

Is the idea not this: for one "velt type" / "memory slice" we have detected a misalignment, and will have cyclic dependency. So vectorization on that "velt type" / "memory slice" should be reverted, for all packs that we already have.
And then, we check if we just rejected the "velt type" / "memory slice" of the current best alignment. If so, we must find a new one.

I am very new to this code, so I might be getting this wrong.

eme64 commented:
I have an example that fails before and after your fix.

Reduced2.java
```
import jdk.internal.misc.Unsafe;

public class Reduced2 {
    static int N = 50;
    static int gold[] = new int[N];

    static Unsafe unsafe = Unsafe.getUnsafe();

    public static void main(String[] strArr) {
        init(gold);
        test(gold);
        for (int i = 0; i < 10_000; i++){
            int[] data = new int[N];
            init(data);
            test(data);
            verify(data, gold);
        }
    }

    static void test(int[] data) {
        for (int i = 2; i < N-2; i++) {
            int v = data[i];
            unsafe.putFloat(data, unsafe.ARRAY_BYTE_BASE_OFFSET + 4 * i + 8, v + 5);
        }
    }

    static void init(int[] data) {
        for (int j = 0; j < N; j++) {
            data[j] = j;
        }
    }

    static void verify(int[] data, int[] gold) {
        for (int i = 0; i < N; i++) {
            if (data[i] != gold[i]) {
                throw new RuntimeException(" Invalid result: dataI[" + i + "]: " + data[i] + " != " + gold[i]);
            }
        }
    }
}
```

Launch it with:
`./java -XX:-TieredCompilation -Xbatch --add-modules java.base --add-exports java.base/jdk.internal.misc=ALL-UNNAMED Reduced2.java`
The assert hits. Add `-Xint` to the command line and it passes.

This happens with your patch:
`StoreF` is looked at first -> `best_align_to_mem_ref`.
Then we look at `StoreI`. We see a misalignment with `StoreF`, so we reject the `StoreI`.
We then go on to remove all packs that have the same type (`int`).
The `StoreF` is never removed.
Eventually, in `best_align_to_mem_ref`, the packs extend `use -> def` all the way back up, via `StoreF <- ConvI2F <- AddI <- LoadI`.

rwestrel commented:
/contributor add eme64

rwestrel commented:
/contributor add epeter

eme64 commented on src/hotspot/share/opto/superword.cpp:
```diff
@@ -662,25 +662,25 @@ void SuperWord::find_adjacent_refs() {
+        // same memory slice since it could be overlapped accesses to the same array.
         create_pack = false;
       } else {
         // Allow independent (different type) unaligned memory operations
```
```suggestion
        // Allow independent (different memory slice) unaligned memory operations
```

eme64 reviewed (changes requested):
Why are you not changing it here?
https://github.com/openjdk/jdk/blob/6751978122afc5102db84c7f040a0b0ab76e70b6/src/hotspot/share/opto/superword.cpp#L728-L739

If we just removed the memory slice where `best_align_to_mem_ref` was in, then should we not find a new `best`?

I fear there is now a way to trigger the assert:
https://github.com/openjdk/jdk/blob/6751978122afc5102db84c7f040a0b0ab76e70b6/src/hotspot/share/opto/superword.cpp#L736
How can we be sure that none of the other packs do not contain a memop with the same type as `mem_ref`?

rwestrel commented:
> Why are you not changing it here?
> 
> https://github.com/openjdk/jdk/blob/6751978122afc5102db84c7f040a0b0ab76e70b6/src/hotspot/share/opto/superword.cpp#L728-L739
> 
> If we just removed the memory slice where `best_align_to_mem_ref` was in, then should we not find a new `best`?

I didn't because for alignment heuristics what matters AFAIU is operand size so the same memory slice check doesn't seem correct. But you're right that there's a problem if `best_align_to_mem_ref` is not part of any pack.

> I fear there is now a way to trigger the assert:

Yes there must be.

I find the code quite confusing because it mixes correctness issues with best alignment heuristics. @vnkozlov I saw in the history that you made the changes that are discussed here. What do you think of this change?

eme64 commented:
@rwestrel I think we can refactor this code in SuperWord.
I hope this change will also go into that direction: [link](https://github.com/openjdk/jdk/pull/12350/files)

rwestrel commented:
Maybe we should first collect all legal packs and then try to find the best alignment, dropping packs that are unaligned on those platforms that support it.

eme64 commented:
But you are right we could refactor the whole code, and separate the steps. That might make things a bit clearer and easier to maintain. But maybe we can do that in a RFE after this fix, and the one for https://github.com/openjdk/jdk/pull/12350

eme64 commented:
I played a bit around, trying to find an example to trigger the assert. It is not very easy.

I found another related problem though (which may be the reason why I cannot find that assert-trigger-example for you now).
https://github.com/openjdk/jdk/blob/6751978122afc5102db84c7f040a0b0ab76e70b6/src/hotspot/share/opto/superword.cpp#L1268-L1271

https://github.com/openjdk/jdk/blob/6751978122afc5102db84c7f040a0b0ab76e70b6/src/hotspot/share/opto/superword.cpp#L629-L641

`isomorphic` would allow memops from different slices to be considered for alignment at the same time. I think we could an additional restriction that only allows memops to be isomorphic if they are also from the same slice.

For this example, I see that the `LoadF` for the `char` and the `byte` array are processed at the same time, for `mem_ref` `466  LoadF  === 500 503 467  [[ 465 ]]  @byte[int:>=0]`.

So far I have not yet been able to reproduce a bug, but I think it should be possible.

```
./java -XX:-TieredCompilation -Xbatch --add-modules java.base --add-exports java.base/jdk.internal.misc=ALL-UNNAMED -XX:CompileCommand=compileonly,Test::test -XX:CompileCommand=printcompilation,Test::test -XX:+TraceSuperWord Test.java
```

```
import jdk.internal.misc.Unsafe;

public class Test {
    static int N = 100;
    static int M = 100*8;
    static byte goldB[] = new byte[M];
    static char goldC[] = new char[M];

    static Unsafe unsafe = Unsafe.getUnsafe();

    public static void main(String[] strArr) {
        init(goldB,goldC);
        test(goldB,goldC);
        for (int i = 0; i < 10_000; i++){
            byte[] dataB = new byte[M];
            char[] dataC = new char[M];
            init(dataB,dataC);
            test(dataB,dataC);
            verify(dataB, goldB);
            verify(dataC, goldC);
        }
    }

    static void test(byte[] dataB, char[] dataC) {
        for (int i = 2; i < N-2; i++) {
            float w = unsafe.getFloat(dataC, unsafe.ARRAY_CHAR_BASE_OFFSET + i*4);
            unsafe.putFloat(dataC, unsafe.ARRAY_CHAR_BASE_OFFSET + i*4, w + 5);
            float v = unsafe.getFloat(dataB, unsafe.ARRAY_BYTE_BASE_OFFSET + i*4);
            unsafe.putInt(dataB, unsafe.ARRAY_BYTE_BASE_OFFSET + i*4, (int)(v + 5));
        }
    }

    static void init(byte[] dataB, char[] dataC) {
        for (int j = 0; j < M; j++) {
            dataB[j] = (byte)j;
            dataC[j] = (char)j;
        }
    }

    static void verify(byte[] data, byte[] gold) {
        for (int i = 0; i < M; i++) {
            if (data[i] != gold[i]) {
                throw new RuntimeException(" Invalid result: data[" + i + "]: " + data[i] + " != " + gold[i]);
            }
        }
    }

    static void verify(char[] data, char[] gold) {
        for (int i = 0; i < M; i++) {
            if (data[i] != gold[i]) {
                throw new RuntimeException(" Invalid result: data[" + i + "]: " + data[i] + " != " + gold[i]);
            }
        }
    }
}
```

I will look into this again once we are done fixing this bug here, and mine in https://github.com/openjdk/jdk/pull/12350.
I think we should be able to break things like this: We have `StoreF` for two different memory slices. They get aligned to the same memref, and pass for this one. But the second one that we smuggle along actually has a conflict on its own slice. But we only do the checks for the first slice. Currently, I cannot find a reproducer without triggering either my bug or the one here at the same time.

So I am noting this here so that I can remember later what to investigate.

vnkozlov commented:
Yes, this code come from JDK-7119644 which improved superword and also take into account SPARC where we did not have unaligned memory vectors. From changes review:
```
New code is added to vectorize operations which have different basic types:

  static void test_IBci(int[] a, byte[] b) {
    for (int i = 0; i < a.length; i+=1) {
      a[i] = -456;
      b[i] = -(byte)123;
    }
  }

0a0   B12: #    B12 B13 <- B11 B12     Loop: B12-B12 inner main of N90 Freq: 9340.37
0a0       movslq  R10, R11    # i2l
0a3       movdqu  [RSI + #16 + R10 << #2],XMM0    ! store vector (16 bytes)
0aa       movq    [RDX + #16 + R10],XMM1    ! store vector (8 bytes)
0b1       movslq  R10, R11    # i2l
0b4       movdqu  [RSI + #32 + R10 << #2],XMM0    ! store vector (16 bytes)
0bb       addl    R11, #8    # int
0bf       cmpl    R11, R8
0c2       jl,s   B12    # loop end  P=0.999893 C=46701.000000

previous code vectorized only one type in such case. New code collects during one iteration
only related memory operations (as before these changes). Then it removes these operations 
from memops list and tries to collect other related mem ops. Such vectors need different loop 
index alignment since vector sizes are different. The new code set maximum loop index alignment. 
Max alignment also works for smaller sizes since sizes are power of 2.
```

vnkozlov commented:
And follow up:
```
> How does it deal with index variables that might be offset?  Something like this:
>
>>  static void test_IBci(int[] a, byte[] b) {
>>    for (int i = 0; i < a.length - 1; i+=1) {
>>      a[i] = -456;
>>      b[i + 1] = -(byte)123;
>>    }
>>  }
>
>
> It's not obvious to me where that will be weeded out.

0b3       movq    [RDX + #17 + R10],XMM1    ! store vector (8 bytes)

It generates unaligned move. For x86 it does not matter (I used only unaligned 
asm instructions) but for SPARC it is disaster (40 times slow since it traps):

IBci: 2382
IBvi: 65

I will need to add a check (current align vs max align) into find_adjacent_refs to 
vectorize only aligned mem ops on SPARC. I want to keep unaligned mem ops 
on x86 since we can win on vector arithmetic.
```

vnkozlov commented:
And:
```
> I found an other problem. My new regression test TestShortVect failed in the overlapping case 
> (load and store from the same array with offset). Superword code thinks store into short[] array 
> is char store and different from short load from short[] so it allows overlapping. It happened 
> because C2 has only one StoreC node for 2 bytes stores but different LoadS and LoadUS 
> nodes for loads. And these load nodes has different memory type: T_SHORT and T_CHAR.

I found that I can't use memory_type() because some ideal transformations (in mulnode.cpp) 
may replace load from char[] with short load and reverse. So instead of comparing vector types 
I added new method same_velt_type(n1, n2) which compare element size for integer types. 
Also there is no difference in vector instructions for Char and Short types so I removed Char 
related vector nodes and used Short vectors for Char type.
```

vnkozlov commented:
> // Can't allow vectorization of unaligned memory accesses with the
> // same type since it could be overlapped accesses to the same array.

It is very simplified protection from cyclic dependency copy when the same array could be passed as different arguments 
as in your test method `testByteByte1()`.

vnkozlov reviewed (commented):
I am not sure that it is correct to relax the check to only bailout for misaligned access to the same array. Unless I forgot how alias indexes created. Do we have the same alias index or different for 2 arrays in `testByteByte1()` test?

eme64 commented:
> Do we have the same alias index or different for 2 arrays in testByteByte1() test?

@vnkozlov as explained in https://github.com/openjdk/jdk/pull/12440#discussion_r1116622180, SuperWord seems to have all arrays of the same basic type in the same slice. If a method has two input arrays, they must be in the same slice, because we do not know if it is maybe the same array.

However, if the two arrays have different basic types, then they are in separate memory slices.

Basically, @rwestrel has detected a subtle difference between `memory slice` and `velt_type`. Usually, the two are identical. But with `Unsafe`, the type of the load/store can be different to that of the array.

@rwestrel : can you change the name of the Bug to reflect this? Suggestion:
`C2: SuperWord alignment analysis must be based on memory slice, not velt type`

vnkozlov commented:
@eme64 thank you for answering my question about alias index.
And I agree we you and Roland that we can add cyclic dependency tests after you fix your bug. 
File RFE for new testing so we don't forget.

eme64 commented:
Hi @rwestrel @vnkozlov .

I have another general question. Should this really vectoroze?
```
    @Test
    @IR(counts = { IRNode.LOAD_VECTOR, ">=1", IRNode.STORE_VECTOR, ">=1" })
    public static void testOffHeapLong1(long dest, long[] src) {
        for (int i = 0; i < src.length; i++) {
            UNSAFE.putLongUnaligned(null, dest + 8 * i, src[i]);
        }
    }
```

I talked with @TobiHartmann about this. It is apparently possible to get the address of an array, and store in a `long`. In that case, we could play with that rawptr a bit, and create a cyclic dependency.

Pseudocode:
```
long[] arr = new long[1000];
long ptr = unsafe.getTheArrayAddress(arr); // does not exist directly, but it is possible
ptr += 8; // shift it one long forward
testOffHeapLong1(ptr, arr);
```

This should have different behavior if it is vectorized or not.

But maybe this is expected, and just a bad use of `Unsafe`. Probably we want this to vectorize, for the extra performance, and people who use `Unsafe` just have to be careful not to do such strange things.

rwestrel commented:
Thanks for the comments @eme64 @vnkozlov 
Looking at the code again, if `vectors_should_be_aligned()` is true, if `create_pack` is false, the current code removes every memops and already created packset with `same_velt_type()` true: That can't be motivated by a correctness issue. So I suppose we want to preserve that behavior. Wouldn't we need the change of the last commit I pushed then?

rwestrel commented:
> But maybe this is expected, and just a bad use of `Unsafe`. Probably we want this to vectorize, for the extra performance, and people who use `Unsafe` just have to be careful not to do such strange things.

Yes, I would say that qualifies as bad use of `Unsafe` and we can't support everything that's possible with it.

eme64 commented:
> Thanks for the comments @eme64 @vnkozlov Looking at the code again, if `vectors_should_be_aligned()` is true, if `create_pack` is false, the current code removes every memops and already created packset with `same_velt_type()` true: That can't be motivated by a correctness issue. So I suppose we want to preserve that behavior. Wouldn't we need the change of the last commit I pushed then?

I think the reason we used `same_velt_type` was that we were confused. Or maybe we did that before we had memory slices, and using `same_velt_type` was at least already an improvemnt? At any rate: it was confused and leads to Bugs in conjunction with `Unsafe`, as my example showed.

Keeping `same_velt_type` will probably not harm much, but be more restrictive than neccessary.
It will not harm much because `velt_type == memory_slice` as long as we are not using `Unsafe`.
And when we do use `Unsafe`, we probably do not use it in very wild ways.

One "wild" use might be something like this:
```
void test(int[] iarr, float[] farr) {
    // cyclic dependency -> not vectorized
    in v1 = (int)Unsafe.LoadF(iarr, i);  // assume this to be best
    Unsafe.StoreI(iarr, i + 1);
    // separate slice -> could be vectorized
    Unsafe.StoreI(farr, i) = Unsafe.LoadI(farr, i);   // on different slice as best, but have same velt_type -> rejected
    // We end up vectorizing nothing, even though we could vectorize the farr
}
```

rwestrel commented:
> Keeping `same_velt_type` will probably not harm much, but be more restrictive than neccessary.

It's quite possible that it's over conservative. What this change is trying to achieve is to relax checks so a pattern that's known to be used in the core libraries optimizes better. That pattern only optimizes for misaligned accesses. So it does seem wrong that those architectures that don't allow misaligned accesses are affected. Also, this code is complicated so it certainly feels safer to me to be on the safe side even if it feels too restrictive. Going forward refactoring this would be nice. I gave it a quick try and it was more complicated than I expected.

I also don't think we should spend too much time making sure every possible combinations of unsafe accesses optimize well or even  correctly if it's too much work. Once people start using unsafe, they are on their own. I think we should stick with whatever feels reasonable or is used in the core libraries (hopefully the second category is included in the first category).

What do you think @vnkozlov ?

eme64 commented:
@rwestrel that sounds ok to me.
BTW I am refactoring the code around your changes, because multiple bugs https://github.com/openjdk/jdk/pull/12350.
I think the code was so badly structured that more and more bugs snuck in.

vnkozlov commented:
> I think the reason we used `same_velt_type` was that we were confused. Or maybe we did that before we had memory slices, and using `same_velt_type` was at least already an improvemnt? At any rate: it was confused and leads to Bugs in conjunction with `Unsafe`, as my example showed.

I did not consider using memory slices (or unsafe access) when worked on this code. Same element type was easy choice for this check.

> I also don't think we should spend too much time making sure every possible combinations of unsafe accesses optimize well or even correctly if it's too much work. Once people start using unsafe, they are on their own. I think we should stick with whatever feels reasonable or is used in the core libraries (hopefully the second category is included in the first category).

Yes, even without vectorization we can construct a Java test with Unsafe access which has cyclic dependencies and overwrite elements intentionally or by mistake. I remember one such case in system libraries several years ago which was fixed.

JIT optimization should not introduce wrong behavior  if Java code does not have it. But if we can correctly detect and reject cyclic dependency we can vectorize it. Original Roland's example and last Emanuel's `StoreI(farr, i)` example don't have "bad" cyclic dependency - at worst they store the same values to the same elements. So it is all about cyclic dependency detection and assumption that we may accessing the same array.

rwestrel commented:
I pushed some more tests than don't vectorize.

rwestrel commented:
> @rwestrel : can you change the name of the Bug to reflect this? Suggestion: `C2: SuperWord alignment analysis must be based on memory slice, not velt type`

The bug database is not used only by developers that are familiar with the code. Sometimes people look for a bug matching what they observe (an error message for instance). So I'm not sure it's a good idea to change the bug synopsis that way as it's meaningless to anyone not familiar with the c2 code.

eme64 commented:
I think `testByteByte2` gets filtered out by the `memory_alignment` check.
https://github.com/openjdk/jdk/blob/f6820c45e9648b2e4f0bf1b5529458e3b3ff3cc5/src/hotspot/share/opto/superword.cpp#L646
or here
https://github.com/openjdk/jdk/blob/f6820c45e9648b2e4f0bf1b5529458e3b3ff3cc5/src/hotspot/share/opto/superword.cpp#L683-L684

That is to be expected. Otherwise you need to turn on `_do_vector_loop`.

rwestrel commented:
At this point, is there something missing from the change so that it can move forward? I'm fine with waiting for #12350 to integrate and then rebase.

rwestrel commented:
> I guess that `testByteByte2` gets filtered out by the `memory_alignment` check.

Are you suggesting, `testByteByte2` should have an IR rule that checks it doesn't vectorize?
It wouldn't be illegal to vectorize while the others are either illegal to vectorize (they wouldn't execute correctly) or they can't be proven to be legal to vectorize with static information.

eme64 commented:
> Are you suggesting, testByteByte2 should have an IR rule that checks it doesn't vectorize?

No, I am not suggesting that. I would not add any IR rule.

But what I would suggest: Add value verification to the tests.
Because we may make mistakes with the vectorization and create wrong results.

eme64 commented:
For me it is also ok if you integrate before https://github.com/openjdk/jdk/pull/12350.

vnkozlov reviewed (approved):
I think we discussed this enough. Approved.

eme64 commented:
Ok. I am running the testing, include stress testing again for commit 10. Will report back on monday.

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java:
```diff
@@ -0,0 +1,239 @@
+        baseOffset = 1;
+        testByteByte5(byteArray, byteArray, 0, size-1);
+    }
+    
```
```suggestion

```

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java:
```diff
@@ -0,0 +1,239 @@
+    public static void testByteByte5_runner() {
+        baseOffset = 1;
+        testByteByte5(byteArray, byteArray, 0, size-1);
+    }
```
```suggestion
    }

```

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java:
```diff
@@ -0,0 +1,239 @@
+        baseOffset = 1;
+        testByteByte5(byteArray, byteArray, 0, size-1);
+    }
+    
```
```suggestion

```

eme64 replied:
Remove the trailing spaces

eme64 reviewed (approved):
Testing passes for change 10. Looks good now.

rwestrel commented:
Thanks for the reviews, testing and discussion @vnkozlov @eme64 
I updated the test case once more:
- Some of the ByteByte test were broken
-  I made sure the test only runs on aarch64 and x64, the platforms I can test this on. Also the test is now skipped if `AlignVector` is false
- I added verification code for the content of the arrays

rwestrel commented:
/integrate

dougxc commented on test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMismatchedAccess.java:
```diff
@@ -0,0 +1,339 @@
+ * @test
+ * @bug 8300258
+ * @key randomness
+ * @requires (os.simpleArch == "x64") | (os.simpleArch == "aarch64")
```
Does this require `@requires compiler.c2`? What is the expectation if this (or any IR) test runs on GraalVM?
cc @chhagedorn

dougxc commented:
Ok, thanks for the explanation. As long as the tests' correctness do not depend on C2, then they will pass on GraalVM, even if they are really testing much in that case.
